proposals/p2360.md

```diff
@@ -0,0 +1,312 @@
+    `i32.(DebugPrintable.Print)()` will look for `i32 as DebugPrintable` not
+    `Type as DebugPrintable`.
+
+## Background
+
+-   [#989: Member access expressions](/proposals/p0989.md) introduced the
+    current compound member access rules.
+-   [Issue #495](https://github.com/carbon-language/carbon-lang/issues/495)
+    suggested adding an implicit conversion from tuples of types to type `Type`.
+-   [Issue #508](https://github.com/carbon-language/carbon-lang/issues/508)
+    raised the question of whether we want tuples like `((), (), ())` to be
+    self-typed.
+
+## Proposal
+
+Instead of treating type-like values as being types regardless of the type of
+those values, we define a _type_ as being a value of type `Type`. Therefore:
+
+-   `()` is a value of type `() as Type`, not a value of type `()` (because the
+    latter is not a type).
+    -   `() as Type` is a value of type `Type`.
+    -   Similarly, `(i32, i32)` is of type `(Type, Type) as Type`, which is of
+        type `Type`.
+    -   Similarly, `{}` is of type `{} as Type`, which is of type `Type`.
+    -   More generally, _the type of any type is `Type`_.
+-   Given `T:! Hashable`, `T` is formally not a type, because it is not of type
+    `Type`. Indeed, member lookup into `Type`, and hence into types, finds
+    nothing, but member lookup into `T` finds members of `Hashable`.
+-   Contexts in which a type is required, such as on the right hand side of a
+    `:` or `:!` binding or the return type of a function, perform an implicit
+    conversion to type `Type`.
+    -   This means that a user-defined type can implement `ImplicitAs(Type)`,
+        annotated in whatever way we eventually decide to annotate functions
+        that are usable at compile time, and values of that type will be usable
+        as types, in the same way that `()` and `{}` are.
+-   There are no singleton type-of-types. It's still possible to have, for
+    example, the type of `i32` be a type whose only value is `i32`. The value
+    `i32` would then not formally be a type, as its type would not be `Type`,
+    but it could still implicitly convert to `Type`, as `()` does, if we so
+    desire.
+    -   This extends to other cases: we could say that for integer literals,
+        which we have recently been describing as having type `IntLiteral(n)`,
+        there is an `impl IntLiteral(n) as ImplicitAs(Type)` which converts a
+        value of type `IntLiteral(n)` to the `Type` value `IntLiteral(n)`. This
+        would then permit `let v: 0 = 0;`.
+    -   Such changes to `i32` or integer literals are not part of this proposal.
+
+## Details
+
+### Terminology
+
+The changes to the language rules in this proposal are fairly small, but there
+is a significant change in how we describe those rules. We now use the following
+terminology:
+
+-   A _type_ is a value of type `Type`.
+-   A _kind_ is a constrained or unconstrained version of `Type`, such as an
```

Comment:
I'm struggling to fully understand this definition, because I don't fully understand what it means for something to be a "version of `Type`". If we're going to repurpose an existing type-theoretic term like "kind" for this, I think it really needs to have a clear definition.